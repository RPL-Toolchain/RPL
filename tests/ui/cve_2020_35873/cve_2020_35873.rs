
struct Session<'a> {
    sess: *mut ffi::Session,
    _f: &'a (),
}

fn str_to_cstring(s: &str) -> Result<std::ffi::CString, Error> {
    std::ffi::CString::new(s).map_err(|_| Error)
}

struct Error;

macro_rules! check {
    ($expr:expr) => {{
        let ret = $expr;
        if ret != 0 {
            return Err(Error);
        }
    }};
}

impl Session<'_> {
    //#[rpl::dump_mir(dump_cfg, dump_ddg)]
    pub fn attach(&mut self, table: Option<&str>) -> Result<(), Error> {
        let table = if let Some(table) = table {
            #[allow(dangling_pointers_from_temporaries)]
            str_to_cstring(table)?.as_ptr()
        } else {
            //~^ NOTE: the `std::ffi::CString` value is dropped here
            std::ptr::null()
        };
        unsafe { check!(ffi::sqlite3session_attach(self.sess, table)) };
        //~^ ERROR: use a pointer from `std::ffi::CString` after dropped
        //~| NOTE: `#[deny(rpl::use_after_drop)]` on by default
        Ok(())
    }
}

mod ffi {
    pub type Session = std::ffi::c_void;
    extern "C" {
        pub fn sqlite3session_attach(s: *mut Session, table: *const std::ffi::c_char) -> i32;
    }
}

fn main() {}