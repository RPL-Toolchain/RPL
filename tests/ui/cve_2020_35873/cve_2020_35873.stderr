error: use a pointer from `std::ffi::CString` after dropped
  --> tests/ui/cve_2020_35873/cve_2020_35873.rs:32:25
   |
LL |         unsafe { check!(ffi::sqlite3session_attach(self.sess, table)) };
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the `std::ffi::CString` value is dropped here
  --> tests/ui/cve_2020_35873/cve_2020_35873.rs:28:9
   |
LL |         } else {
   |         ^

error: aborting due to 1 previous error

